packages/million/block.ts

```diff
@@ -0,0 +1,261 @@
+/* eslint-disable @typescript-eslint/unbound-method */
+import {
+  childNodes$,
+  cloneNode$,
+  createEventListener,
+  insertBefore$,
+  insertText,
+  innerHTML$,
+  remove$ as removeElement$,
+  setAttribute,
+  setText,
+  setStyleAttribute,
+  setSvgAttribute,
+  firstChild$,
+  nextSibling$,
+  weakMapHas$,
+  weakMapGet$,
+  weakMapSet$,
+} from './dom';
+import { renderToTemplate } from './template';
+import { AbstractBlock } from './types';
+import type { Edit, EditChild, Props, VElement, Hole, VNode } from './types';
+
+export const createBlock = (
+  fn: (props?: Props) => VElement,
+  unwrap?: (vnode: any) => VNode,
+) => {
+  const vnode = fn(
+    new Proxy(
+      {},
+      {
+        // A universal getter will return a Hole instance if props[any] is accessed
+        // Allows code to identify holes in virtual nodes ("digs" them out)
+        get(_, key: string): Hole {
+          return { __key: key };
+        },
+      },
+    ),
+  );
+  const edits: Edit[] = [];
+
+  // Turns vnode into a string of HTML and creates an array of "edits"
+  // Edits are instructions for how to update the DOM given some props
+  const root = stringToDOM(
+    renderToTemplate(unwrap ? unwrap(vnode) : vnode, edits),
+  );
+
+  // Handles case for positioning text nodes. When text nodes are
+  // put into a template, they can be merged. For example,
+  // ["hello", "world"] becomes "helloworld" in the DOM.
+  // Inserts text nodes into the DOM at the correct position.
+  for (let i = 0, j = edits.length; i < j; ++i) {
+    const current = edits[i]!;
+    const initsLength = current.inits.length;
+    if (!initsLength) continue;
+    const el = getCurrentElement(current, root);
+    for (let k = 0; k < initsLength; ++k) {
+      const init = current.inits[k]!;
+      insertText(el, init.value, init.index);
+    }
+  }
+
+  return (
+    props?: Props | null,
+    key?: string,
+    shouldUpdate?: (oldProps: Props, newProps: Props) => boolean,
+  ) => {
+    return new Block(root, edits, props, key ?? props?.key, shouldUpdate);
+  };
+};
+
+export class Block extends AbstractBlock {
+  root: HTMLElement;
+  edits: Edit[];
+  // Cache for getCurrentElement()
+  cache = new WeakMap<Edit, HTMLElement>();
```

Comment:
I just realize why `cache` is never getting released.

Imagine this case:

```ts
const block = createBlock();

const row1 = block({}); // This initiates a new Block instance, along with a `.cache`
const row2 = block({}); // This also initiates a new Block instance. Now we have 2 Block instances
// ...
const row1000 = block({}); // Now we have 1000 Block instances
```

If those rows never get GC-ed, so do those 1000 Block instances, and so do those 1000 caches.

Reply:
Yeah, edits should and will never get GC'd because createBlock acts almost like creating a "component." That's why I thought the numeric index may have been better, but I don't think Maps can be GC'd which means the memory usage is roughly the same

Reply:
> That's why I thought the numeric index may have been better

We might need a benchmark on this.

Since `edits` will never get GC-ed, the `Map` should be preferred over `WeakMAap`, as `WeakMap` is definitely slower.

Reply:
benchmarks have roughly the same results